adapter-static: Error when building site under Linux

### Describe the bug

I’ve got an SK project I started back when it was at v1.0.0-next.107. It essentially consists of a single route which preloads data from a single endpoint. I just finished it and it works in dev mode, so I added adapter-static. Now I can build it on Windows with <kbd>svelte-kit build</kbd>, with the final output being:
```
> Using @sveltejs/adapter-static
  ✔ done
```
And <kbd>svelte-kit preview</kbd> works correctly.

However, <kbd>svelte-kit build</kbd> fails under (Alpine) Linux with an error about trying to open a directory as a file. This means I can’t deploy it anywhere.

Since I started with a much older version, it could just be that I need to update my code to use a new style, but I can’t see any problems based on the docs. I created an MWE using <kbd>npm init svelte@next</kbd>. (My project uses TypeScript, but I was able to reproduce it with a JavaScript skeleton project.) I’m attaching the link and the logs in the relevant fields. Here’s the file tree, just for completeness:
```
.
├── .gitignore
├── _data
│  └── foo.yaml
├── jsconfig.json
├── package-lock.json
├── package.json
├── README.md
├── src
│  ├── app.html
│  ├── global.d.ts
│  └── routes
│     ├── data.js
│     └── index.svelte
├── static
│  └── favicon.png
└── svelte.config.js
```

### Reproduction

https://github.com/shivjm/sveltekit-static-test

### Logs

```
> sveltekit-test@0.0.1 build /work.new
> svelte-kit build

vite v2.6.14 building for production...
✓ 14 modules transformed.
.svelte-kit/output/client/_app/manifest.json                    1.21 KiB
.svelte-kit/output/client/_app/layout.svelte-712392a1.js        0.53 KiB / gzip: 0.35 KiB
.svelte-kit/output/client/_app/error.svelte-55ba95a6.js         1.56 KiB / gzip: 0.74 KiB
.svelte-kit/output/client/_app/pages/index.svelte-db8ac5cc.js   0.74 KiB / gzip: 0.47 KiB
.svelte-kit/output/client/_app/start-ebb02e83.js                18.38 KiB / gzip: 6.80 KiB
.svelte-kit/output/client/_app/assets/start-61d1577b.css        0.16 KiB / gzip: 0.15 KiB
.svelte-kit/output/client/_app/chunks/vendor-af0f2ef7.js        7.08 KiB / gzip: 2.89 KiB
vite v2.6.14 building SSR bundle for production...
✓ 15 modules transformed.
.svelte-kit/output/server/app.js                      0.07 KiB
.svelte-kit/output/server/chunks/app-498ddc9a.js      47.98 KiB
.svelte-kit/output/server/chunks/data-e92d2162.js     0.20 KiB
.svelte-kit/output/server/chunks/layout-dc4a3a6e.js   0.23 KiB
.svelte-kit/output/server/chunks/error-31498736.js    0.71 KiB
.svelte-kit/output/server/chunks/index-3bbafa22.js    0.67 KiB

Run npm run preview to preview your production build locally.

> Using @sveltejs/adapter-static
> EISDIR: illegal operation on a directory, open 'build/data/'
Error: EISDIR: illegal operation on a directory, open 'build/data/'
    at Object.openSync (fs.js:497:3)
    at writeFileSync (fs.js:1528:35)
    at file:///work.new/node_modules/@sveltejs/kit/dist/chunks/index4.js:234:6
    at Map.forEach (<anonymous>)
    at visit (file:///work.new/node_modules/@sveltejs/kit/dist/chunks/index4.js:220:17)
    at async prerender (file:///work.new/node_modules/@sveltejs/kit/dist/chunks/index4.js:298:6)
    at async Object.prerender (file:///work.new/node_modules/@sveltejs/kit/dist/chunks/index4.js:363:4)
    at async adapt (file:///work.new/node_modules/@sveltejs/adapter-static/index.js:13:4)
    at async adapt (file:///work.new/node_modules/@sveltejs/kit/dist/chunks/index4.js:388:2)
    at async file:///work.new/node_modules/@sveltejs/kit/dist/cli.js:891:5
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! sveltekit-test@0.0.1 build: `svelte-kit build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the sveltekit-test@0.0.1 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-11-14T10_41_13_461Z-debug.log
```


### System Info

```shell
System:
    OS: Linux 5.10 Alpine Linux
    CPU: (16) x64 AMD Ryzen 7 2700 Eight-Core Processor
    Memory: 48.15 GB / 50.15 GB
    Container: Yes
    Shell: 1.31.1 - /bin/ash
  Binaries:
    Node: 14.18.1 - /usr/local/bin/node
    Yarn: 1.22.15 - /usr/local/bin/yarn
    npm: 6.14.15 - /usr/local/bin/npm
```


### Severity

blocking all usage of SvelteKit

### Additional Information

_No response_

--- a/.changeset/tender-geckos-agree.md
+++ b/.changeset/tender-geckos-agree.md
@@ -1,0 +1,5 @@
+---
+'@sveltejs/kit': patch
+---
+
+Never allow trailing slash for endpoints

--- a/.changeset/tender-geckos-agree.md
+++ b/.changeset/tender-geckos-agree.md
@@ -2,4 +2,4 @@
 '@sveltejs/kit': patch
 ---
 
-Ignore `trailingSlash` for endpoint requests, apply `trailingSlash` to pages consistently
+[breaking] Ignore `trailingSlash` for endpoint requests, apply `trailingSlash` to pages consistently

--- a/.changeset/tender-geckos-agree.md
+++ b/.changeset/tender-geckos-agree.md
@@ -2,4 +2,4 @@
 '@sveltejs/kit': patch
 ---
 
-Never allow trailing slash for endpoints
+Ignore `trailingSlash` for endpoint requests, apply `trailingSlash` to pages consistently

--- a/documentation/docs/14-configuration.md
+++ b/documentation/docs/14-configuration.md
@@ -275,7 +275,7 @@ An object containing zero or more of the following values:
 
 ### trailingSlash
 
-Whether to remove, append, or ignore trailing slashes when resolving URLs to routes.
+Whether to remove, append, or ignore trailing slashes when resolving URLs (note that this only applies to pages, not endpoints).
 
 - `'never'` — redirect `/x/` to `/x`
 - `'always'` — redirect `/x` to `/x/`

--- a/packages/kit/src/core/preview/index.js
+++ b/packages/kit/src/core/preview/index.js
@@ -7,7 +7,6 @@ import { pathToFileURL } from 'url';
 import { getRequest, setResponse } from '../../node.js';
 import { installFetch } from '../../install-fetch.js';
 import { SVELTE_KIT_ASSETS } from '../constants.js';
-import { normalize_path } from '../../utils/url.js';
 
 /** @typedef {import('http').IncomingMessage} Req */
 /** @typedef {import('http').ServerResponse} Res */
@@ -90,7 +89,7 @@ export async function preview({ port, host, config, https: use_https = false }) 
 				return;
 			}
 
-			const { pathname, search } = new URL(/** @type {string} */ (req.url), 'http://dummy');
+			const { pathname } = new URL(/** @type {string} */ (req.url), 'http://dummy');
 
 			// only treat this as a page if it doesn't include an extension
 			if (pathname === '/' || /\/[^./]+\/?$/.test(pathname)) {

--- a/packages/kit/src/core/preview/index.js
+++ b/packages/kit/src/core/preview/index.js
@@ -92,16 +92,6 @@ export async function preview({ port, host, config, https: use_https = false }) 
 
 			const { pathname, search } = new URL(/** @type {string} */ (req.url), 'http://dummy');
 
-			const normalized = normalize_path(pathname, config.kit.trailingSlash);
-
-			if (normalized !== pathname) {
-				res.writeHead(307, {
-					location: base + normalized + search
-				});
-				res.end();
-				return;
-			}
-
 			// only treat this as a page if it doesn't include an extension
 			if (pathname === '/' || /\/[^./]+\/?$/.test(pathname)) {
 				const file = join(

--- a/packages/kit/src/runtime/server/index.js
+++ b/packages/kit/src/runtime/server/index.js
@@ -85,11 +85,8 @@ export async function respond(request, options, state) {
 		}
 	}
 
-	if (route) {
-		const normalized = normalize_path(
-			url.pathname,
-			route.type === 'endpoint' ? 'never' : options.trailing_slash
-		);
+	if (route?.type === 'page') {
+		const normalized = normalize_path(url.pathname, options.trailing_slash);
 
 		if (normalized !== url.pathname && !state.prerender?.fallback) {
 			return new Response(undefined, {

--- a/packages/kit/src/runtime/server/index.js
+++ b/packages/kit/src/runtime/server/index.js
@@ -16,20 +16,6 @@ const default_transform = ({ html }) => html;
 export async function respond(request, options, state) {
 	let url = new URL(request.url);
 
-	const normalized = normalize_path(url.pathname, options.trailing_slash);
-
-	if (normalized !== url.pathname && !state.prerender?.fallback) {
-		return new Response(undefined, {
-			status: 301,
-			headers: {
-				location:
-					// ensure paths starting with '//' are not treated as protocol-relative
-					(normalized.startsWith('//') ? url.origin + normalized : normalized) +
-					(url.search === '?' ? '' : url.search)
-			}
-		});
-	}
-
 	const { parameter, allowed } = options.method_override;
 	const method_override = url.searchParams.get(parameter)?.toUpperCase();
 
@@ -96,6 +82,25 @@ export async function respond(request, options, state) {
 				params = decode_params(matched);
 				break;
 			}
+		}
+	}
+
+	if (route) {
+		const normalized = normalize_path(
+			url.pathname,
+			route.type === 'endpoint' ? 'never' : options.trailing_slash
+		);
+
+		if (normalized !== url.pathname && !state.prerender?.fallback) {
+			return new Response(undefined, {
+				status: 301,
+				headers: {
+					location:
+						// ensure paths starting with '//' are not treated as protocol-relative
+						(normalized.startsWith('//') ? url.origin + normalized : normalized) +
+						(url.search === '?' ? '' : url.search)
+				}
+			});
 		}
 	}
 

--- a/packages/kit/src/utils/url.js
+++ b/packages/kit/src/utils/url.js
@@ -46,7 +46,7 @@ export function normalize_path(path, trailing_slash) {
 
 	if (trailing_slash === 'never') {
 		return path.endsWith('/') ? path.slice(0, -1) : path;
-	} else if (trailing_slash === 'always' && /\/[^./]+$/.test(path)) {
+	} else if (trailing_slash === 'always' && !path.endsWith('/')) {
 		return path + '/';
 	}
 

--- a/packages/kit/src/utils/url.spec.js
+++ b/packages/kit/src/utils/url.spec.js
@@ -67,11 +67,6 @@ test('normalizes paths', () => {
 			ignore: '/foo/',
 			always: '/foo/',
 			never: '/foo'
-		},
-		'/foo.json': {
-			ignore: '/foo.json',
-			always: '/foo.json',
-			never: '/foo.json'
 		}
 	};
 

--- a/packages/kit/test/apps/options/source/pages/endpoint.js
+++ b/packages/kit/test/apps/options/source/pages/endpoint.js
@@ -1,5 +1,5 @@
 export function get() {
 	return {
-		body: 'this should be accessible as /path-base/endpoint, not /path-base/endpoint/'
+		body: 'hi'
 	};
 }

--- a/packages/kit/test/apps/options/source/pages/endpoint.js
+++ b/packages/kit/test/apps/options/source/pages/endpoint.js
@@ -1,0 +1,5 @@
+export function get() {
+	return {
+		body: 'this should be accessible as /path-base/endpoint, not /path-base/endpoint/'
+	};
+}

--- a/packages/kit/test/apps/options/test/test.js
+++ b/packages/kit/test/apps/options/test/test.js
@@ -143,6 +143,14 @@ test.describe.parallel('trailingSlash', () => {
 		expect(page.url()).toBe(`${baseURL}/path-base/slash/child/`);
 		expect(await page.textContent('h2')).toBe('/slash/child/');
 	});
+
+	test('does not add trailing slash to endpoint', async ({ baseURL, request }) => {
+		const response = await request.get('/path-base/endpoint/');
+		expect(response.url()).toBe(`${baseURL}/path-base/endpoint`);
+		expect(await response.text()).toBe(
+			'this should be accessible as /path-base/endpoint, not /path-base/endpoint/'
+		);
+	});
 });
 
 test.describe.parallel('serviceWorker', () => {

--- a/packages/kit/test/apps/options/test/test.js
+++ b/packages/kit/test/apps/options/test/test.js
@@ -144,12 +144,14 @@ test.describe.parallel('trailingSlash', () => {
 		expect(await page.textContent('h2')).toBe('/slash/child/');
 	});
 
-	test('does not add trailing slash to endpoint', async ({ baseURL, request }) => {
-		const response = await request.get('/path-base/endpoint/');
-		expect(response.url()).toBe(`${baseURL}/path-base/endpoint`);
-		expect(await response.text()).toBe(
-			'this should be accessible as /path-base/endpoint, not /path-base/endpoint/'
-		);
+	test('ignores trailing slash on endpoint', async ({ baseURL, request }) => {
+		const r1 = await request.get('/path-base/endpoint/');
+		expect(r1.url()).toBe(`${baseURL}/path-base/endpoint/`);
+		expect(await r1.text()).toBe('hi');
+
+		const r2 = await request.get('/path-base/endpoint');
+		expect(r2.url()).toBe(`${baseURL}/path-base/endpoint`);
+		expect(await r2.text()).toBe('hi');
 	});
 });
 

